bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/SecretRepository.cs

```diff
@@ -252,4 +243,70 @@ public async Task<IEnumerable<Core.SecretsManager.Entities.Secret>> ImportAsync(
         }
         return secrets;
     }
+
+    public async Task<(bool Read, bool Write)> AccessToSecretAsync(Guid id, Guid userId, AccessClientType accessType)
+    {
+        using var scope = ServiceScopeFactory.CreateScope();
+        var dbContext = GetDatabaseContext(scope);
+
+        var accessPolicy = await dbContext.Secret
+            .Where(s => s.Id == id)
+            .Select(s => new
+            {
+                Read = s.Projects.Any(p =>
+                    p.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Read) ||
+                    p.GroupAccessPolicies.Any(ap =>
+                        ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Read))),
+                Write = s.Projects.Any(p =>
+                    p.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Write) ||
+                    p.GroupAccessPolicies.Any(ap =>
+                        ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Write))),
+            }).FirstOrDefaultAsync();
+
+        return (accessPolicy.Read, accessPolicy.Write);
+    }
```

Comment:
I really like the interface for this and the concept. 

If I understand things correctly, we will need to handle different cases in the access client type?

Something like 

```
    public async Task<(bool Read, bool Write)> AccessToSecretAsync(Guid id, Guid userId, AccessClientType accessType)
    {
        using var scope = ServiceScopeFactory.CreateScope();
        var dbContext = GetDatabaseContext(scope);

        var accessPolicyQuery = dbContext.Secret
            .Where(s => s.Id == id);

        switch (accessType)
        {
            case AccessClientType.NoAccessCheck:
                return (true, true);
            case AccessClientType.User:
            {
                var ap = await accessPolicyQuery.Select(s => new
                {
                    Read = s.Projects.Any(p =>
                        p.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Read) ||
                        p.GroupAccessPolicies.Any(ap =>
                            ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Read))),
                    Write = s.Projects.Any(p =>
                        p.UserAccessPolicies.Any(ap => ap.OrganizationUser.User.Id == userId && ap.Write) ||
                        p.GroupAccessPolicies.Any(ap =>
                            ap.Group.GroupUsers.Any(gu => gu.OrganizationUser.User.Id == userId && ap.Write))),
                }).FirstOrDefaultAsync();
                return (ap.Read, ap.Write);
            }
            case AccessClientType.ServiceAccount:
            {
                var ap = await accessPolicyQuery.Select(s => new
                {
                    Read = s.Projects.Any(p =>
                        p.ServiceAccountAccessPolicies.Any(ap => ap.ServiceAccountId == userId && ap.Read)), 
                    Write = false, // Change when SAs can have write access
                }).FirstOrDefaultAsync();
                return (ap.Read, ap.Write);
            }
            default:
                return (false, false); // OR throw an exception?
        }
    }
```

Reply:
Yes,